core/src/main/scala/kafka/coordinator/transaction/TransactionStateManager.scala

```diff
@@ -204,7 +217,7 @@ class TransactionStateManager(brokerId: Int,
 
         try {
           while (currOffset < logEndOffset
-            && loadingPartitions.contains(topicPartition.partition())
+            && inLock(stateLock) {loadingPartitions.exists(_.txnPartitionId == topicPartition.partition)}
```

Comment:
It would be useful to check if the coordinator epoch is still the expected here as well.